syntax = "proto3";

service Jacaranda {
  rpc compile(WasmFunction) returns (NativeBinary) {}
}

message FunctionType {
  uint32 param_count = 1;
  // Will be of type char, see LanguageTypes in wasm.h
  // Encoded as varint size so four byte width doesn't matter
  repeated uint32 param_types = 2;
  uint32 return_count = 3;
  // Defaults to 0 if not present
  optional uint32 return_type = 4;
}

message WasmFunction {
  // A hash of the module:function:architecture pair
  string hash = 1;
  uint32 func_idx = 2;
  bool is_imported = 3;
  FunctionType func_type = 4;
  bytes func_body = 5;
  string target_data_layout = 6;
  uint32 program_pointer_size = 7;
}

message NativeBinary {
  bytes data_bytes = 1;
  uint32 data_length = 2;
}